### PR TITLE
feat: Adds GlobalClusterSelfManagedSharding to Cluster

### PR DESCRIPTION
**Pull request by EspenAlbert**
## Proposed changes

Adds GlobalClusterSelfManagedSharding to Cluster and a new example: `cluster-self-managed-sharding.json`

Link to any related issue(s): CLOUDP-256267

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as
  expected)
- [ ] This change requires a documentation update
- [ ] If changes include removal or addition of 3rd party GitHub actions, I updated our internal document. Reach out to the APIx Integration slack channel to get access to the internal document.

## Manual QA performed:

- [x] cfn invoke for each of CRUDL/cfn test
- [x] Updated resource in  [example](https://github.com/mongodb/mongodbatlas-cloudformation-resources/tree/master/examples)
- [x] Published to AWS private registry
- [x] Used the template in [example](https://github.com/mongodb/mongodbatlas-cloudformation-resources/tree/master/examples) to create and update a stack in AWS
- [x] Deleted stack to ensure resources are deleted
- [ ] Created multiple resources in same stack
- [x] Validated in Atlas UI
- [x] Included screenshots

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run `make fmt` and formatted my code
- [x] For CFN Resources: I have released by changes in the private registry and proved by change
  works in Atlas

## Further comments

Screenshots:

Atlas UI of example:

image

contract tests:

image

CFN stack (see [Jira Ticket](https://jira.mongodb.org/browse/CLOUDP-256534) for fixing updates to replication spec):

image

### CODE REVIEW

File: cfn-resources/cluster/mongodb-atlas-cluster.json
Comment (lantoli) [inline cfn-resources/cluster/mongodb-atlas-cluster.json:345]: description is fine but I'm working in adding a little bit more of detail about how to configure, it should be fast, i'll keep you posted.
Please wait for merge, or alternatively you can merge and do the doc changes in another PR

File: cfn-resources/cluster/cmd/resource/mappings.go
Comment (lantoli) [reply cfn-resources/cluster/cmd/resource/mappings.go:42]: i have mixed feelings here. On the one hand I think it's a good idea to have contract tests about as many features and resources as possible.
on the other hand this feature requires GEOSHARDED clusters that are very slow to create, update, etc., and we'll be probably testing more Atlas than the CFN logic, so in this case I'm not sure about the pros/cons balance

**Review by lantoli — APPROVED** [review]

File: cfn-resources/cluster/mongodb-atlas-cluster.json
Comment (lantoli) [reply cfn-resources/cluster/mongodb-atlas-cluster.json:345]: @EspenAlbert I have a proposal doc in this PR: https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2348
WDYT about waiting until doc team reviews?

Comment (EspenAlbert) [reply cfn-resources/cluster/mongodb-atlas-cluster.json:345]: Sounds good 👍

File: cfn-resources/cluster/cmd/resource/mappings.go
Comment (maastha) [reply cfn-resources/cluster/cmd/resource/mappings.go:42]: I understand but I'm not sure if that's a good enough reason to not have a test at all. Looking at the current [inputs](https://github.com/mongodb/mongodbatlas-cloudformation-resources/blob/master/cfn-resources/cluster/test/inputs_1_create.template.json) file for cluster CFN resource, it looks like there is very minimal coverage as it is and only one REPLICASET test case with only one replication spec is being tested.

Also I don't see cluster tests running in CI.

I'd suggest we should have a test to avoid further divergence from the resource. Since the test doesn't run in CI anyway it would be good to at least have the option to run it locally.

Comment (lantoli) [reply cfn-resources/cluster/cmd/resource/mappings.go:42]: if there are not e2e or contract tests cluster tests we should definitely have.

my point was that we should have only some basic tests in e2e / contract tests, not all different combinations of parameters and features, that can normally better tested in unit tests.

in this case I'm not sure about having a specific e2e or contract test only to test this attribute, but I agree to have some e2e and contract tests running in CI for cluster.

File: cfn-resources/cluster/mongodb-atlas-cluster.json
Comment (lantoli) [reply cfn-resources/cluster/mongodb-atlas-cluster.json:345]: doc is already reviewed in that PR, you can already take the doc from there

Comment (EspenAlbert) [reply cfn-resources/cluster/mongodb-atlas-cluster.json]: Nice. Updated in [781c02c](https://github.com/mongodb/mongodbatlas-cloudformation-resources/pull/1079/commits/781c02ce2c66d3646210a28804c9791a539a0b37)
Need a restamp

File: cfn-resources/cluster/cmd/resource/mappings.go
Comment (EspenAlbert) [reply cfn-resources/cluster/cmd/resource/mappings.go:42]: Added a follow-up ticket [here](https://jira.mongodb.org/browse/CLOUDP-256730)

**Review by lantoli — DISMISSED** [review]

File: cfn-resources/cluster/mongodb-atlas-cluster.json
Comment (lantoli) [inline cfn-resources/cluster/mongodb-atlas-cluster.json:345]: as this is the only doc, I'll probably go for the long explanation (the one in the TF resource)

File: cfn-resources/cluster/docs/README.md
Comment (lantoli) [inline cfn-resources/cluster/docs/README.md:146]: same, as this is the only doc, I'll probably go for the long explanation (the one in the TF resource)

File: cfn-resources/cluster/mongodb-atlas-cluster.json
Comment (AgustinBettati) [inline cfn-resources/cluster/mongodb-atlas-cluster.json:454]: q: Does defining this property as createOnly have any impact when `GlobalClusterSelfManagedSharding` is left empty and then populated with value coming from API?

File: cfn-resources/cluster/docs/README.md
Comment (AgustinBettati) [inline cfn-resources/cluster/docs/README.md:152]: similar to the concern mentioned in terraform, it would seem that marking property as `createOnly` forces a replacement if modified. Is this intentional? We could consider leaving as a regular property. cc: @lantoli

File: cfn-resources/cluster/mongodb-atlas-cluster.json
Comment (EspenAlbert) [reply cfn-resources/cluster/mongodb-atlas-cluster.json:454]: The CFN implementation differs a bit from Terraform:
1. If the value is `nil` then we use the value from the API
2. else, we use the value specified in the model

The normal test case/examples don't specify this flag and still managed the full CRUD operations, so I believe we should be safe 👍

File: cfn-resources/cluster/docs/README.md
Comment (EspenAlbert) [reply cfn-resources/cluster/docs/README.md:152]: See [comment above](https://github.com/mongodb/mongodbatlas-cloudformation-resources/pull/1079#discussion_r1645675545)

Comment (lantoli) [reply cfn-resources/cluster/docs/README.md:152]: @EspenAlbert I think it's useful to do a manual test of creating a stack and then modifying the param and update the stack to see if we get an error (correct) instead of a cluster replacement (bad)

File: cfn-resources/cluster/cmd/resource/mappings.go
Comment (maastha) [reply cfn-resources/cluster/cmd/resource/mappings.go:42]: >not all different combinations of parameters and features, that can be normally better tested in unit tests.

I'm not suggesting to have tests for all combinations of parameters, but at least for different types of cluster, I would consider that basic coverage. 
In this case, the type would be GEOSHARDED using this attribute.

File: cfn-resources/cluster/docs/README.md
Comment (EspenAlbert) [reply cfn-resources/cluster/docs/README.md:152]: image

I switched it from `null` to `false` and got the error message above (no cluster deletion) ☝️ 

Then I did a different update (setting `PitEnabled: false` (was true)) and it worked (see screenshot 👇 )

image

Can you confirm this is the correct behavior @lantoli ?

Comment (marcosuma) [reply cfn-resources/cluster/docs/README.md:152]: Good catch, I wonder if that works as expected or it's an unexpected API behavior? 

Also @EspenAlbert any chance to convert this into a test instead of leaving it manual?

Comment (EspenAlbert) [reply cfn-resources/cluster/docs/README.md:152]: @marcosuma 
To me it is expected behavior ✅  but I want to check with @lantoli before merging.
Yes, plan on adding that in a separate ticket, see [discussion above](https://github.com/mongodb/mongodbatlas-cloudformation-resources/pull/1079#discussion_r1645469457)

Comment (lantoli) [reply cfn-resources/cluster/docs/README.md:152]: @EspenAlbert thanks, can you do the last test of starting with attribute true, and then try to change to false and if it fails or the cluster is recreated?

Comment (lantoli) [reply cfn-resources/cluster/docs/README.md:152]: @EspenAlbert just to be completely sure (in case null is treated differently), can you do 2 more tests, one creating with false and trying to change to true, and the opposite, starting with true and change to false? thanks!

**Review by lantoli — APPROVED** [review]

File: cfn-resources/cluster/docs/README.md
Comment (EspenAlbert) [reply cfn-resources/cluster/docs/README.md:152]: Done ✅ 
Getting the same error in both cases

image

**Pull request merged**
